File: src/main/java/net/nerfatg/proxy/packet/Packet.java
package net.nerfatg.proxy.packet;

import java.nio.BufferOverflowException;
import java.nio.ByteBuffer;

public abstract class Packet {

    private PacketType type;
    private PacketAction action;

    public Packet(ByteBuffer buffer) throws BufferOverflowException {
        this.fromBytes(buffer);
    }

    public Packet(PacketType type, PacketAction action) {
        this.type = type;
        this.action = action;
    }

    public PacketType getType() {
        return type;
    }

    public PacketAction getAction() {
        return action;
    }

    public void fromBytes(ByteBuffer buffer) {
        this.type = PacketType.values()[buffer.get(0)];
        this.action = PacketAction.values()[buffer.get(1)];
        this.readPayload(buffer, 4); // Payload begins after 4 bytes
    }

    public void toBytes(ByteBuffer buffer) {
        buffer.put(0, (byte) type.ordinal());
        buffer.put(1, (byte) action.ordinal());
        this.writePayload(buffer, 4); // Payload begins after 4 bytes
    }

    public abstract void readPayload(ByteBuffer buffer, int offset) throws BufferOverflowException;
    public abstract void writePayload(ByteBuffer buffer, int offset) throws BufferOverflowException;
}

File: Packets/Packet.cs


using System;
using Unity.VisualScripting;

public abstract class Packet<T> where T : Enum
{
    public T Type { get; private set; }
    public PacketAction Action { get; set; }
    public Packet(byte[] bytes)
    {
        this.Type = (T)Enum.ToObject(typeof(T), bytes[0]);
        this.Action = (PacketAction)bytes[1];
        this.FromBytes(bytes);
    }

    public Packet(T type, PacketAction action)
    {
        this.Type = type;
        this.Action = action;
    }

    public void ToBytes(byte[] bytes)
    {
        bytes[0] = Convert.ToByte(Type);
        bytes[1] = (byte)Action;
        WritePayload(bytes, 4); // Payload begins after 4 bytes
    }

    public void FromBytes(byte[] bytes)
    {
        ReadPayload(bytes, 4);
    }

    protected abstract void ReadPayload(byte[] bytes, byte offset);
    protected abstract void WritePayload(byte[] bytes, byte offset);
}

public enum PacketAction
{
    Generic,
    Add,
    Update,
    Remove,
    Replace
}

